llvm/lib/CAS/PluginCAS.cpp

```diff
@@ -240,12 +244,51 @@ Expected<ObjectHandle> PluginObjectStore::load(ObjectRef Ref) {
   case LLCAS_LOOKUP_RESULT_SUCCESS:
     return makeObjectHandle(c_obj.opaque);
   case LLCAS_LOOKUP_RESULT_NOTFOUND:
-    report_fatal_error("PluginCAS: object reference not found");
+    return std::nullopt;
   case LLCAS_LOOKUP_RESULT_ERROR:
     return Ctx->errorAndDispose(c_err);
   }
 }
 
+void PluginObjectStore::loadAsync(
+    ObjectRef Ref,
+    unique_function<void(Expected<std::optional<ObjectHandle>>)> Callback) {
+  llcas_objectid_t c_id{Ref.getInternalRef(*this)};
+
+  struct LoadObjCtx {
+    std::shared_ptr<PluginCASContext> CASCtx;
+    PluginObjectStore &CAS;
```

Comment:
Now that we are having async actions, I wonder if we should change `PluginObjectStore` to have `shared_from_this` and use a `shared_ptr` or `weak_ptr` here so we cannot have UAF in these callbacks.

Reply:
Note: this could be a follow-up, doesn't need to block this.

Reply:
I've done this (https://github.com/apple/llvm-project/pull/6458/commits/c48add9fc9f05a41f85e2c0470b7c8a7c11d70fe) for `PluginObjectStore` but similar concern exists for `ObjectStore::getProxyAsync` and that one seems more tricky. I added a `FIXME` for that function.